src/IIIFAuth2/IIIFAuth2.API/Infrastructure/Auth/CustomerDomainService.cs

```diff
@@ -0,0 +1,118 @@
+﻿using IIIFAuth2.API.Data;
+using IIIFAuth2.API.Utils;
+
+namespace IIIFAuth2.API.Infrastructure.Auth;
+
+public interface ICustomerDomainProvider
+{
+    /// <summary>
+    /// Get a list of any custom domains configured for customer
+    /// </summary>
+    Task<IReadOnlyCollection<string>> GetCustomerCookieDomains(int customerId);
+}
+
+public interface ICustomerDomainChecker
+{
+    /// <summary>
+    /// Check if the specified Origin is for the current domain, or it is on a domain that the DLCS can issue cookies to
+    /// </summary>
+    Task<bool> OriginForControlledDomain(int customerId, Uri origin);
+}
+
+/// <summary>
+/// Service to help dealing with customer domains
+/// </summary>
+public class CustomerDomainService : ICustomerDomainProvider, ICustomerDomainChecker
+{
+    private readonly IHttpContextAccessor httpContextAccessor;
+    private readonly AuthServicesContext dbContext;
+    private readonly ILogger<CustomerDomainService> logger;
+
+    public CustomerDomainService(
+        IHttpContextAccessor httpContextAccessor,
+        AuthServicesContext dbContext,
+        ILogger<CustomerDomainService> logger)
+    {
+        this.httpContextAccessor = httpContextAccessor;
+        this.dbContext = dbContext;
+        this.logger = logger;
+    }
+    
+    /// <summary>
+    /// Check if the specified Origin is for the current domain, or it is on a domain that the DLCS can issue cookies to
+    /// </summary>
+    public async Task<bool> OriginForControlledDomain(int customerId, Uri origin)
+    {
+        try
+        {
+            var currentRequest = httpContextAccessor.SafeHttpContext().Request;
+            if (OriginMatchesHost(origin, currentRequest)) return true;
+            if (OriginSubdomainOfHost(origin, currentRequest)) return true;
+
+            return await IsOriginSubdomainOfCookieDomain(customerId, origin);
+        }
+        catch (Exception ex)
+        {
+            logger.LogError(ex, "Error checking if Origin is for ControlledDomain");
+            return true;
```

Comment:
would this be better to return something other than true? To me this looks like there's been some sort of error, so we don't know whether the origin is for the current domain

Reply:
I think a boolean is fine as a return value, even in the event of an error. This method checks whether we need a significant gesture or we can safely issue a cookie, if an error occurs I err on the side of caution and require one which seemed better than showing the user an error (slightly annoying ux vs broken). The only way that I can see an error occurring is when hitting the database.

Returning `true` is incorrect though! It should be `false` as a fallback to require a significant gesture.